File: 6.Software-Design-Patterns/example/Creational/Prototype/Prototype.go
package main

import "fmt"

// UserAddress represents the address information for a user
type UserAddress struct {
  City    string `json:"city"`
  Country string `json:"country"`
}

// User represents a user with name and address
type User struct {
  Name     string     `json:"name"`
  Address  UserAddress `json:"address"`
}

// NewUser creates a new User object
func NewUser(name string, address UserAddress) *User {
  return &User{Name: name, Address: address}
}

// Clone creates a deep copy of the User object
func (u *User) Clone() *User {
  // Create a new User object with a copy of the address
  return NewUser(u.Name, UserAddress{City: u.Address.City, Country: u.Address.Country})
}

func main() {
  // Create a prototype user
  prototypeUser := NewUser("Prototype User", UserAddress{City: "Prototype City", Country: "Prototype Country"})

  // Clone the prototype user
  clonedUser := prototypeUser.Clone()

  fmt.Println("Prototype User:", prototypeUser)
  fmt.Println("Cloned User:", clonedUser)

  // Modify the cloned user
  clonedUser.Name = "Cloned User"
  clonedUser.Address.City = "Cloned City"

  fmt.Println("Prototype User (after modification):", prototypeUser)
  fmt.Println("Cloned User (after modification):", clonedUser)
}

File: 6.Software-Design-Patterns/example/Creational/Prototype/Prototype.java
// Package declaration for organization (replace with your package name)
package com.example.prototype;

// Interface for user addresses with clear naming convention
public interface UserAddress {

  // Getter methods for city and country
  String getCity();
  String getCountry();
}

// Immutable class representing a User address
public final class UserAddressImpl implements UserAddress {

  private final String city;
  private final String country;

  public UserAddressImpl(String city, String country) {
    this.city = city;
    this.country = country;
  }

  @Override
  public String getCity() {
    return city;
  }

  @Override
  public String getCountry() {
    return country;
  }
}

// Prototype class representing a User with Javadoc comments
/**
 * Represents a User with name and address information.
 * This class serves as a prototype for creating new User objects.
 */
public class UserPrototype {

  private final String name;
  private final UserAddress address;

  /**
   * Constructor for UserPrototype.
   *
   * @param name User's name
   * @param address User's address
   */
  public UserPrototype(String name, UserAddress address) {
    this.name = name;
    this.address = address;
  }

  /**
   * Creates a deep copy of the UserPrototype object.
   *
   * @return A new UserPrototype object with the same data as the current one.
   */
  public UserPrototype clone() {
    // Create a new UserPrototype with the same name and a deep copy of the address
    return new UserPrototype(name, new UserAddressImpl(address.getCity(), address.getCountry()));
  }

  @Override
  public String toString() {
    return "UserPrototype{" +
        "name='" + name + '\'' +
        ", address=" + address +
        '}';
  }
}

// Concrete class inheriting from the UserPrototype
public class User extends UserPrototype {

  public User(String name, UserAddress address) {
    super(name, address);
  }
}

// Separate class for program entry point
public class PrototypeDemo {

  public static void main(String[] args) {
    UserAddress prototypeAddress = new UserAddressImpl("Prototype City", "Prototype Country");
    UserPrototype prototypeUser = new UserPrototype("Prototype User", prototypeAddress);

    UserPrototype clonedUser = prototypeUser.clone();

    System.out.println("Prototype User: " + prototypeUser);
    System.out.println("Cloned User: " + clonedUser);

    // Modifying the cloned user
    clonedUser.setName("Cloned User");
    ((UserAddressImpl) clonedUser.getAddress()).setCity("Cloned City"); // Casting for modification (avoid if possible)

    System.out.println("Prototype User (after modification): " + prototypeUser);
    System.out.println("Cloned User (after modification): " + clonedUser);
  }
}
